"""
DO NOT EDIT THIS FILE!!
This file is automatically generated by the process_schemas.py program
in the scripts directory. It is not intended to be edited directly. If
you need to update the GEL protocol classes, please run the script
on the appropriate schema version.
"""
from protocols.protocol import ProtocolElement
from protocols.protocol import SearchRequest
from protocols.protocol import SearchResponse

import avro.schema

version = '0.2.0'


class AdoptedStatus(object):
    """
    Default is not_adopted adoptedin means adopted into the family
    adoptedout means child belonged to the family and was adopted out
    """
    not_adopted = "not_adopted"
    adoptedin = "adoptedin"
    adoptedout = "adoptedout"


class CalledGenotype(ProtocolElement):
    """
    This is intended to hold the genotypes for the family members In
    principle it is a phased zygosity as in VCF spec and called by the
    analysis provider if further phasing is conducted
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"CalledGenotype", "fields": [{"doc": "", "type": "string", "name":
"participantId"}, {"doc": "", "type": "string", "name": "genotype"},
{"doc": "", "type": ["null", "float"], "name": "alleleFraction"},
{"doc": "", "type": ["null", "int"], "name": "copyNumber"}], "doc":
""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "alleleFraction",
        "copyNumber",
        "genotype",
        "participantId",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'alleleFraction', 'copyNumber', 'genotype', 'participantId'
    ]

    def __init__(self, **kwargs):
        self.alleleFraction = kwargs.get(
            'alleleFraction', None)
        self.copyNumber = kwargs.get(
            'copyNumber', None)
        self.genotype = kwargs.get(
            'genotype', None)
        self.participantId = kwargs.get(
            'participantId', None)


class CalledGenotypes(ProtocolElement):
    """
    This is intended to hold the genotypes for the family members In
    principle it is a phased zygosity as in VCF spec and called by the
    analysis provider if further phasing is conducted
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"CalledGenotypes", "fields": [{"doc": "", "type": "string", "name":
"sampleId"}, {"doc": "", "type": "string", "name": "genotype"},
{"type": ["null", "float"], "name": "alleleFraction"}, {"doc": "",
"type": ["null", "int"], "name": "copyNumber"}], "doc": ""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "alleleFraction",
        "copyNumber",
        "genotype",
        "sampleId",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'alleleFraction', 'copyNumber', 'genotype', 'sampleId'
    ]

    def __init__(self, **kwargs):
        self.alleleFraction = kwargs.get(
            'alleleFraction', None)
        self.copyNumber = kwargs.get(
            'copyNumber', None)
        self.genotype = kwargs.get(
            'genotype', None)
        self.sampleId = kwargs.get(
            'sampleId', None)


class CancerDemographics(ProtocolElement):
    """
    This defines a Cancer Demographics
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"CancerDemographics", "fields": [{"type": "string", "name": "id"},
{"type": ["null", {"items": "string", "type": "array"}], "name":
"externalIds"}, {"type": ["null", "string"], "name":
"originatingCenter"}, {"doc": "", "type": {"symbols": ["pilot",
"main"], "doc": "", "type": "enum", "name": "ProgramPhase"}, "name":
"programPhase"}, {"type": ["null", "string"], "name":
"centerPatientId"}, {"doc": "", "type": ["null", {"doc": "", "type":
"record", "name": "CancerRecruitmentCategory", "fields": [{"type":
"string", "name": "cancerName"}, {"type": "string", "name":
"dataModelVersion"}]}], "name": "cancerRecruitmentCategory"}, {"doc":
"", "type": ["null", "string"], "name": "primaryDiagnosis"}, {"doc":
"", "type": {"symbols": ["male", "female", "unknown"], "doc": "",
"type": "enum", "name": "Sex"}, "name": "sex"}, {"type": {"fields":
[{"doc": "", "type": "ProgramPhase", "name": "programPhase"},
{"default": false, "doc": "", "type": "boolean", "name": "optOut"},
{"default": false, "type": "boolean", "name": "programmeConsent"},
{"default": false, "doc": "", "type": "boolean", "name":
"primaryFindingConsent"}, {"default": false, "doc": "", "type":
"boolean", "name": "secondaryFindingConsent"}, {"default": false,
"doc": "", "type": "boolean", "name": "carrierStatusConsent"}],
"type": "record", "name": "ConsentStatus"}, "name": "consentStatus"},
{"doc": "", "type": ["null", {"values": "string", "type": "map"}],
"name": "additionalInformation"}], "doc": ""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "additionalInformation",
        "cancerRecruitmentCategory",
        "centerPatientId",
        "consentStatus",
        "externalIds",
        "id",
        "originatingCenter",
        "primaryDiagnosis",
        "programPhase",
        "sex",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'cancerRecruitmentCategory': CancerRecruitmentCategory,
            'consentStatus': ConsentStatus,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'cancerRecruitmentCategory': CancerRecruitmentCategory,
            'consentStatus': ConsentStatus,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'additionalInformation', 'cancerRecruitmentCategory',
        'centerPatientId', 'consentStatus', 'externalIds', 'id',
        'originatingCenter', 'primaryDiagnosis', 'programPhase', 'sex'
    ]

    def __init__(self, **kwargs):
        self.additionalInformation = kwargs.get(
            'additionalInformation', None)
        self.cancerRecruitmentCategory = kwargs.get(
            'cancerRecruitmentCategory', None)
        self.centerPatientId = kwargs.get(
            'centerPatientId', None)
        self.consentStatus = kwargs.get(
            'consentStatus', None)
        self.externalIds = kwargs.get(
            'externalIds', None)
        self.id = kwargs.get(
            'id', None)
        self.originatingCenter = kwargs.get(
            'originatingCenter', None)
        self.primaryDiagnosis = kwargs.get(
            'primaryDiagnosis', None)
        self.programPhase = kwargs.get(
            'programPhase', None)
        self.sex = kwargs.get(
            'sex', None)


class CancerInterpretationAnalysis(ProtocolElement):
    """
    TODO: change InterpretationAnalysis > CancerInterpretationAnalysis
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"CancerInterpretationAnalysis", "fields": [{"doc": "", "type":
"string", "name": "reportRequestId"}, {"doc": "", "type": "string",
"name": "analysisId"}, {"type": "string", "name": "companyName"},
{"type": "string", "name": "reportURI"}, {"type": {"items": {"fields":
[{"doc": "", "type": "string", "name": "chromosome"}, {"doc": "",
"type": "string", "name": "dbSNPid"}, {"doc": "", "type": "int",
"name": "position"}, {"doc": "", "type": "string", "name":
"reference"}, {"doc": "", "type": "string", "name": "alternate"},
{"doc": "", "type": {"items": {"fields": [{"doc": "", "type": ["null",
"string"], "name": "modeOfInheritance"}, {"doc": "", "type":
{"fields": [{"doc": "", "type": "string", "name": "symbol"}, {"doc":
"", "type": "string", "name": "transcriptId"}, {"doc": "", "type":
{"values": "string", "type": "map"}, "name": "ids"}, {"doc": "",
"type": ["null", "int"], "name": "exomeSize"}, {"doc": "", "type":
["null", "double"], "name": "geneCoverage"}], "type": "record",
"name": "Gene"}, "name": "gene"}, {"doc": "", "type": "float", "name":
"score"}, {"doc": "", "type": ["null", "boolean"], "name":
"fullyExplainsPhenotype"}, {"doc": "", "type": ["null", "int"],
"name": "groupOfVariants"}, {"doc": "", "type": "string", "name":
"variantJustification"}, {"type": ["null", {"symbols": ["NONE",
"TIER1", "TIER2", "TIER3"], "doc": "", "type": "enum", "name":
"Tier"}], "name": "tier"}], "type": "record", "name": "ReportEvents"},
"type": "array"}, "name": "reportEvents"}, {"type": {"items": {"doc":
"", "type": "record", "name": "CalledGenotypes", "fields": [{"doc":
"", "type": "string", "name": "sampleId"}, {"doc": "", "type":
"string", "name": "genotype"}, {"type": ["null", "float"], "name":
"alleleFraction"}, {"doc": "", "type": ["null", "int"], "name":
"copyNumber"}]}, "type": "array"}, "name": "calledGenotypes"},
{"type": ["null", "float"], "name": "internalScore"}, {"type":
["null", {"symbols": ["BENIGN", "LIKELY_BENIGN", "VUS",
"LIKELY_PATHOGENIC", "PATHOGENIC"], "doc": "", "type": "enum", "name":
"VariantClassification"}], "name": "variantClassification"}, {"doc":
"", "type": ["null", {"values": "string", "type": "map"}], "name":
"additionalTextualVariantAnnotations"}, {"doc": "", "type": ["null",
{"values": "float", "type": "map"}], "name":
"additionalNumericVariantAnnotations"}, {"doc": "", "type": ["null",
{"items": "string", "type": "array"}], "name": "comments"}, {"doc":
"", "type": {"symbols": ["unknown", "somatic", "germline"], "doc": "",
"type": "enum", "name": "SomaticOrGermline"}, "name":
"somaticOrGermline"}], "type": "record", "name": "ReportedVariants"},
"type": "array"}, "name": "reportedVariants"}, {"type": {"items":
{"fields": [{"doc": "", "type": "string", "name": "chromosome"},
{"doc": "", "type": "int", "name": "start"}, {"doc": "", "type":
"int", "name": "end"}, {"doc": "", "type": "string", "name": "type"},
{"doc": "", "type": "string", "name": "reference"}, {"doc": "",
"type": "string", "name": "alternate"}, {"type": {"items":
"ReportEvents", "type": "array"}, "name": "ReportEvents"}, {"type":
{"items": "CalledGenotypes", "type": "array"}, "name":
"calledGenotypes"}, {"doc": "", "type": ["null", "float"], "name":
"internalScore"}, {"type": ["null", "VariantClassification"], "name":
"variantClassification"}, {"doc": "", "type": ["null", {"values":
"string", "type": "map"}], "name":
"additionalTextualVariantAnnotations"}, {"doc": "", "type": ["null",
{"values": "float", "type": "map"}], "name":
"additionalNumericVariantAnnotations"}, {"doc": "", "type": ["null",
{"items": "string", "type": "array"}], "name": "comments"}, {"doc":
"", "type": "SomaticOrGermline", "name": "somaticOrGermline"}],
"type": "record", "name": "ReportedStructuralVariants"}, "type":
"array"}, "name": "reportedStructuralVariants"}, {"doc": "", "type":
["null", {"items": "string", "type": "array"}], "name": "comments"}],
"doc": ""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "analysisId",
        "comments",
        "companyName",
        "reportRequestId",
        "reportURI",
        "reportedStructuralVariants",
        "reportedVariants",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'reportedStructuralVariants': ReportedStructuralVariants,
            'reportedVariants': ReportedVariants,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'reportedStructuralVariants': ReportedStructuralVariants,
            'reportedVariants': ReportedVariants,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'analysisId', 'comments', 'companyName', 'reportRequestId',
        'reportURI', 'reportedStructuralVariants', 'reportedVariants'
    ]

    def __init__(self, **kwargs):
        self.analysisId = kwargs.get(
            'analysisId', None)
        self.comments = kwargs.get(
            'comments', None)
        self.companyName = kwargs.get(
            'companyName', None)
        self.reportRequestId = kwargs.get(
            'reportRequestId', None)
        self.reportURI = kwargs.get(
            'reportURI', None)
        self.reportedStructuralVariants = kwargs.get(
            'reportedStructuralVariants', None)
        self.reportedVariants = kwargs.get(
            'reportedVariants', None)


class CancerInterpretationRequest(ProtocolElement):
    """
    This record represents basic information for this report  TODO:
    change/specialise InterpretationRequest >
    CancerInterpretationRequest currently does not load... python
    build.py
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"CancerInterpretationRequest", "fields": [{"doc": "", "type":
"string", "name": "reportRequestId"}, {"doc": "", "type": "int",
"name": "reportVersion"}, {"default": false, "doc": "", "type":
"boolean", "name": "interpretGenome"}, {"doc": "", "type": {"items":
{"doc": "", "type": "record", "name": "File", "fields": [{"doc": "",
"type": ["null", "string", {"items": "string", "type": "array"}],
"name": "SampleId"}, {"doc": "", "type": "string", "name":
"URIFile"}]}, "type": "array"}, "name": "BAMs"}, {"doc": "", "type":
{"items": "File", "type": "array"}, "name": "gVCFs"}, {"doc": "",
"type": {"items": "File", "type": "array"}, "name": "VCFs"}, {"doc":
"", "type": {"items": "File", "type": "array"}, "name": "BigWigs"},
{"doc": "", "type": "File", "name": "annotationFile"}, {"doc": "",
"type": {"doc": "", "type": "record", "name": "VirtualPanels",
"fields": [{"doc": "", "type": "string", "name": "cancerName"},
{"doc": "", "type": "string", "name": "dataModelCatalogueVersion"},
{"doc": "", "type": ["null", {"items": "string", "type": "array"}],
"name": "geneIds"}, {"doc": "", "type": ["null", {"items": "string",
"type": "array"}], "name": "clinicalRelevantTranscripts"}, {"doc": "",
"type": ["null", "File"], "name": "RelevantRegions"}, {"doc": "",
"type": ["null", "File"], "name": "clinicalRelevantVariants"}]},
"name": "virtualPanels"}, {"type": "string", "name": "analysisURI"},
{"type": "string", "name": "analysisVersion"}, {"doc": "", "type":
["null", {"values": "string", "type": "map"}], "name":
"additionalInfo"}, {"doc": "", "type": {"doc": "", "type": "record",
"name": "OtherFamilyHistory", "fields": [{"type": ["null", {"items":
"string", "type": "array"}], "name": "maternalFamilyHistory"},
{"type": ["null", {"items": "string", "type": "array"}], "name":
"paternalFamilyHistory"}]}, "name": "otherFamilyHistory"}], "doc": ""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "BAMs",
        "BigWigs",
        "VCFs",
        "additionalInfo",
        "analysisURI",
        "analysisVersion",
        "annotationFile",
        "gVCFs",
        "otherFamilyHistory",
        "reportRequestId",
        "reportVersion",
        "virtualPanels",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'BAMs': File,
            'BigWigs': File,
            'VCFs': File,
            'annotationFile': File,
            'gVCFs': File,
            'otherFamilyHistory': OtherFamilyHistory,
            'virtualPanels': VirtualPanels,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'BAMs': File,
            'BigWigs': File,
            'VCFs': File,
            'annotationFile': File,
            'gVCFs': File,
            'otherFamilyHistory': OtherFamilyHistory,
            'virtualPanels': VirtualPanels,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'BAMs', 'BigWigs', 'VCFs', 'additionalInfo', 'analysisURI',
        'analysisVersion', 'annotationFile', 'gVCFs',
        'interpretGenome', 'otherFamilyHistory', 'reportRequestId',
        'reportVersion', 'virtualPanels'
    ]

    def __init__(self, **kwargs):
        self.BAMs = kwargs.get(
            'BAMs', None)
        self.BigWigs = kwargs.get(
            'BigWigs', None)
        self.VCFs = kwargs.get(
            'VCFs', None)
        self.additionalInfo = kwargs.get(
            'additionalInfo', None)
        self.analysisURI = kwargs.get(
            'analysisURI', None)
        self.analysisVersion = kwargs.get(
            'analysisVersion', None)
        self.annotationFile = kwargs.get(
            'annotationFile', None)
        self.gVCFs = kwargs.get(
            'gVCFs', None)
        self.interpretGenome = kwargs.get(
            'interpretGenome', False)
        self.otherFamilyHistory = kwargs.get(
            'otherFamilyHistory', None)
        self.reportRequestId = kwargs.get(
            'reportRequestId', None)
        self.reportVersion = kwargs.get(
            'reportVersion', None)
        self.virtualPanels = kwargs.get(
            'virtualPanels', None)


class CancerParticipant(ProtocolElement):
    """
    This defines a Cancer Participant (demographics and sample
    information)
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"CancerParticipant", "fields": [{"type": {"doc": "", "type": "record",
"name": "CancerDemographics", "fields": [{"type": "string", "name":
"id"}, {"type": ["null", {"items": "string", "type": "array"}],
"name": "externalIds"}, {"type": ["null", "string"], "name":
"originatingCenter"}, {"doc": "", "type": {"symbols": ["pilot",
"main"], "doc": "", "type": "enum", "name": "ProgramPhase"}, "name":
"programPhase"}, {"type": ["null", "string"], "name":
"centerPatientId"}, {"doc": "", "type": ["null", {"doc": "", "type":
"record", "name": "CancerRecruitmentCategory", "fields": [{"type":
"string", "name": "cancerName"}, {"type": "string", "name":
"dataModelVersion"}]}], "name": "cancerRecruitmentCategory"}, {"doc":
"", "type": ["null", "string"], "name": "primaryDiagnosis"}, {"doc":
"", "type": {"symbols": ["male", "female", "unknown"], "doc": "",
"type": "enum", "name": "Sex"}, "name": "sex"}, {"type": {"fields":
[{"doc": "", "type": "ProgramPhase", "name": "programPhase"},
{"default": false, "doc": "", "type": "boolean", "name": "optOut"},
{"default": false, "type": "boolean", "name": "programmeConsent"},
{"default": false, "doc": "", "type": "boolean", "name":
"primaryFindingConsent"}, {"default": false, "doc": "", "type":
"boolean", "name": "secondaryFindingConsent"}, {"default": false,
"doc": "", "type": "boolean", "name": "carrierStatusConsent"}],
"type": "record", "name": "ConsentStatus"}, "name": "consentStatus"},
{"doc": "", "type": ["null", {"values": "string", "type": "map"}],
"name": "additionalInformation"}]}, "name": "cancerDemographics"},
{"doc": "", "type": {"items": {"doc": "", "type": "record", "name":
"CancerSample", "fields": [{"doc": "", "type": "string", "name":
"id"}, {"type": ["null", "string"], "name": "centerSampleId"}, {"doc":
"", "type": {"symbols": ["germline", "tumor"], "type": "enum", "name":
"SampleType"}, "name": "sampleType"}, {"doc": "", "type": ["null",
"string"], "name": "source"}, {"doc": "", "type": ["null", {"symbols":
["FFPE", "FF", "Unknown", "LEUK", "GL"], "type": "enum", "name":
"PreservationMethod"}], "name": "preservationMethod"}, {"doc": "",
"type": ["null", {"symbols": ["primary", "metastasis"], "doc": "",
"type": "enum", "name": "Phase"}], "name": "phase"}, {"doc": "",
"type": ["null", {"symbols": ["resection", "biopsy"], "doc": "",
"type": "enum", "name": "Method"}], "name": "method"}, {"doc": "",
"type": ["null", "double"], "name": "cellularity"}, {"doc": "",
"type": ["null", "double"], "name": "tumorContent"}]}, "type":
"array"}, "name": "cancerSamples"}, {"doc": "", "type": {"items":
{"doc": "", "type": "record", "name": "MatchedSamples", "fields":
[{"type": "string", "name": "germlineSampleId"}, {"type": "string",
"name": "tumorSampleId"}]}, "type": "array"}, "name":
"matchedSamples"}], "doc": ""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "cancerDemographics",
        "cancerSamples",
        "matchedSamples",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'cancerDemographics': CancerDemographics,
            'cancerSamples': CancerSample,
            'matchedSamples': MatchedSamples,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'cancerDemographics': CancerDemographics,
            'cancerSamples': CancerSample,
            'matchedSamples': MatchedSamples,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'cancerDemographics', 'cancerSamples', 'matchedSamples'
    ]

    def __init__(self, **kwargs):
        self.cancerDemographics = kwargs.get(
            'cancerDemographics', None)
        self.cancerSamples = kwargs.get(
            'cancerSamples', None)
        self.matchedSamples = kwargs.get(
            'matchedSamples', None)


class CancerRecruitmentCategory(ProtocolElement):
    """
    EIK: added - This records describes the data model the patient was
    recruited under  - important in defining the cancerPanel for
    analysis and interpretation
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"CancerRecruitmentCategory", "fields": [{"type": "string", "name":
"cancerName"}, {"type": "string", "name": "dataModelVersion"}], "doc":
""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "cancerName",
        "dataModelVersion",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'cancerName', 'dataModelVersion'
    ]

    def __init__(self, **kwargs):
        self.cancerName = kwargs.get(
            'cancerName', None)
        self.dataModelVersion = kwargs.get(
            'dataModelVersion', None)


class CancerSample(ProtocolElement):
    """
    stage in some cases it might be nice to know about the grade and
    stage of tumor, but this is not strictly required
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"CancerSample", "fields": [{"doc": "", "type": "string", "name":
"id"}, {"type": ["null", "string"], "name": "centerSampleId"}, {"doc":
"", "type": {"symbols": ["germline", "tumor"], "type": "enum", "name":
"SampleType"}, "name": "sampleType"}, {"doc": "", "type": ["null",
"string"], "name": "source"}, {"doc": "", "type": ["null", {"symbols":
["FFPE", "FF", "Unknown", "LEUK", "GL"], "type": "enum", "name":
"PreservationMethod"}], "name": "preservationMethod"}, {"doc": "",
"type": ["null", {"symbols": ["primary", "metastasis"], "doc": "",
"type": "enum", "name": "Phase"}], "name": "phase"}, {"doc": "",
"type": ["null", {"symbols": ["resection", "biopsy"], "doc": "",
"type": "enum", "name": "Method"}], "name": "method"}, {"doc": "",
"type": ["null", "double"], "name": "cellularity"}, {"doc": "",
"type": ["null", "double"], "name": "tumorContent"}], "doc": ""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "cellularity",
        "centerSampleId",
        "id",
        "method",
        "phase",
        "preservationMethod",
        "sampleType",
        "source",
        "tumorContent",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'cellularity', 'centerSampleId', 'id', 'method', 'phase',
        'preservationMethod', 'sampleType', 'source', 'tumorContent'
    ]

    def __init__(self, **kwargs):
        self.cellularity = kwargs.get(
            'cellularity', None)
        self.centerSampleId = kwargs.get(
            'centerSampleId', None)
        self.id = kwargs.get(
            'id', None)
        self.method = kwargs.get(
            'method', None)
        self.phase = kwargs.get(
            'phase', None)
        self.preservationMethod = kwargs.get(
            'preservationMethod', None)
        self.sampleType = kwargs.get(
            'sampleType', None)
        self.source = kwargs.get(
            'source', None)
        self.tumorContent = kwargs.get(
            'tumorContent', None)


class CarrierStatus(object):
    """
    Definitions of enumeration used in RDParticipant
    """
    unaffected = "unaffected"
    affected = "affected"
    carrier = "carrier"
    uncertain = "uncertain"
    presymptomatic = "presymptomatic"


class ConsentStatus(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"ConsentStatus", "fields": [{"doc": "", "type": {"symbols": ["pilot",
"main"], "doc": "", "type": "enum", "name": "ProgramPhase"}, "name":
"programPhase"}, {"default": false, "doc": "", "type": "boolean",
"name": "optOut"}, {"type": ["null", "string"], "name":
"consentVersion"}, {"default": false, "type": "boolean", "name":
"programmeConsent"}, {"default": false, "doc": "", "type": "boolean",
"name": "primaryFindingConsent"}, {"default": false, "doc": "",
"type": "boolean", "name": "secondaryFindingConsent"}, {"default":
false, "doc": "", "type": "boolean", "name": "carrierStatusConsent"}]}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "consentVersion",
        "programPhase",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'carrierStatusConsent', 'consentVersion', 'optOut',
        'primaryFindingConsent', 'programPhase', 'programmeConsent',
        'secondaryFindingConsent'
    ]

    def __init__(self, **kwargs):
        self.carrierStatusConsent = kwargs.get(
            'carrierStatusConsent', False)
        self.consentVersion = kwargs.get(
            'consentVersion', None)
        self.optOut = kwargs.get(
            'optOut', False)
        self.primaryFindingConsent = kwargs.get(
            'primaryFindingConsent', False)
        self.programPhase = kwargs.get(
            'programPhase', None)
        self.programmeConsent = kwargs.get(
            'programmeConsent', False)
        self.secondaryFindingConsent = kwargs.get(
            'secondaryFindingConsent', False)


class Disorder(ProtocolElement):
    """
    This is quite GEL specific. This is the way is stored in
    ModelCatalogue and PanelApp. More generally one would like to have
    an omim or orphanet id here Currently all specific disease titles
    are assigned to a disease subgroup so really only specificDisease
    needs to be completed but we add the others for generality
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"Disorder", "fields": [{"doc": "", "type": ["null", "string"], "name":
"diseaseGroup"}, {"doc": "", "type": ["null", "string"], "name":
"diseaseSubGroup"}, {"doc": "", "type": ["null", "string"], "name":
"specificDisease"}, {"doc": "", "type": ["null", "int"], "name":
"ageOfOnset"}], "doc": ""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "ageOfOnset",
        "diseaseGroup",
        "diseaseSubGroup",
        "specificDisease",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'ageOfOnset', 'diseaseGroup', 'diseaseSubGroup',
        'specificDisease'
    ]

    def __init__(self, **kwargs):
        self.ageOfOnset = kwargs.get(
            'ageOfOnset', None)
        self.diseaseGroup = kwargs.get(
            'diseaseGroup', None)
        self.diseaseSubGroup = kwargs.get(
            'diseaseSubGroup', None)
        self.specificDisease = kwargs.get(
            'specificDisease', None)


class File(ProtocolElement):
    """
    This defines a file This Record is defined by the sampleID and a
    URI Currently SampleID can be a single String or an array of
    strings if multiple samples are assoicated with the same file
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name": "File",
"fields": [{"doc": "", "type": ["null", "string", {"items": "string",
"type": "array"}], "name": "SampleId"}, {"doc": "", "type": "string",
"name": "URIFile"}], "doc": ""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "SampleId",
        "URIFile",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'SampleId', 'URIFile'
    ]

    def __init__(self, **kwargs):
        self.SampleId = kwargs.get(
            'SampleId', None)
        self.URIFile = kwargs.get(
            'URIFile', None)


class Gene(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name": "Gene",
"fields": [{"doc": "", "type": "string", "name": "symbol"}, {"doc":
"", "type": "string", "name": "transcriptId"}, {"doc": "", "type":
{"values": "string", "type": "map"}, "name": "ids"}, {"doc": "",
"type": ["null", "int"], "name": "exomeSize"}, {"doc": "", "type":
["null", "double"], "name": "geneCoverage"}]}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "exomeSize",
        "geneCoverage",
        "ids",
        "symbol",
        "transcriptId",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'exomeSize', 'geneCoverage', 'ids', 'symbol', 'transcriptId'
    ]

    def __init__(self, **kwargs):
        self.exomeSize = kwargs.get(
            'exomeSize', None)
        self.geneCoverage = kwargs.get(
            'geneCoverage', None)
        self.ids = kwargs.get(
            'ids', None)
        self.symbol = kwargs.get(
            'symbol', None)
        self.transcriptId = kwargs.get(
            'transcriptId', None)


class HpoTerm(ProtocolElement):
    """
    This defines an HPO term and its modifiers (possibly multiple)
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"HpoTerm", "fields": [{"doc": "", "type": "string", "name": "term"},
{"doc": "", "type": ["null", "boolean"], "name": "termPresence"},
{"doc": "", "type": ["null", {"items": "string", "type": "array"}],
"name": "modifier"}, {"doc": "", "type": ["null", "int"], "name":
"ageOfOnset"}], "doc": ""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "ageOfOnset",
        "modifier",
        "term",
        "termPresence",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'ageOfOnset', 'modifier', 'term', 'termPresence'
    ]

    def __init__(self, **kwargs):
        self.ageOfOnset = kwargs.get(
            'ageOfOnset', None)
        self.modifier = kwargs.get(
            'modifier', None)
        self.term = kwargs.get(
            'term', None)
        self.termPresence = kwargs.get(
            'termPresence', None)


class InterpretationAnalysis(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"InterpretationAnalysis", "fields": [{"doc": "", "type": "string",
"name": "reportRequestId"}, {"doc": "", "type": "string", "name":
"analysisId"}, {"type": "string", "name": "companyName"}, {"type":
"string", "name": "reportURI"}, {"type": {"items": {"fields": [{"doc":
"", "type": "string", "name": "chromosome"}, {"doc": "", "type":
"string", "name": "dbSNPid"}, {"doc": "", "type": "int", "name":
"position"}, {"doc": "", "type": "string", "name": "reference"},
{"doc": "", "type": "string", "name": "alternate"}, {"doc": "",
"type": {"items": {"doc": "", "type": "record", "name":
"CalledGenotype", "fields": [{"doc": "", "type": "string", "name":
"participantId"}, {"doc": "", "type": "string", "name": "genotype"},
{"doc": "", "type": ["null", "float"], "name": "alleleFraction"},
{"doc": "", "type": ["null", "int"], "name": "copyNumber"}]}, "type":
"array"}, "name": "calledGenotypes"}, {"doc": "", "type": {"items":
{"fields": [{"doc": "", "type": "string", "name":
"modeOfInheritance"}, {"doc": "", "type": {"fields": [{"doc": "",
"type": "string", "name": "symbol"}, {"doc": "", "type": "string",
"name": "transcriptId"}, {"doc": "", "type": {"values": "string",
"type": "map"}, "name": "ids"}, {"doc": "", "type": ["null", "int"],
"name": "exomeSize"}, {"doc": "", "type": ["null", "float"], "name":
"geneCoverage"}], "type": "record", "name": "Gene"}, "name": "gene"},
{"doc": "", "type": {"symbols": ["complete", "incomplete"], "doc": "",
"type": "enum", "name": "Penetrance"}, "name": "penetrance"}, {"doc":
"", "type": "float", "name": "score"}, {"doc": "", "type": ["null",
"boolean"], "name": "fullyExplainsPhenotype"}, {"doc": "", "type":
["null", "int"], "name": "groupOfVariants"}, {"doc": "", "type":
"string", "name": "variantJustification"}, {"type": ["null",
{"symbols": ["NONE", "TIER1", "TIER2", "TIER3"], "doc": "", "type":
"enum", "name": "Tier"}], "name": "tier"}], "type": "record", "name":
"ReportEvent"}, "type": "array"}, "name": "reportEvents"}, {"doc": "",
"type": ["null", "float"], "name": "internalScore"}, {"type": ["null",
{"symbols": ["BENIGN", "LIKELY_BENIGN", "VUS", "LIKELY_PATHOGENIC",
"PATHOGENIC"], "doc": "", "type": "enum", "name":
"VariantClassification"}], "name": "variantClassification"}, {"doc":
"", "type": ["null", {"values": "string", "type": "map"}], "name":
"additionalTextualVariantAnnotations"}, {"doc": "", "type": ["null",
{"values": "float", "type": "map"}], "name":
"additionalNumericVariantAnnotations"}, {"doc": "", "type": ["null",
{"items": "string", "type": "array"}], "name": "comments"}], "type":
"record", "name": "ReportedVariant"}, "type": "array"}, "name":
"reportedVariants"}, {"type": ["null", {"items": {"fields": [{"doc":
"", "type": "string", "name": "chromosome"}, {"doc": "", "type":
"int", "name": "start"}, {"doc": "", "type": "int", "name": "end"},
{"doc": "", "type": "string", "name": "type"}, {"doc": "", "type":
"string", "name": "reference"}, {"doc": "", "type": "string", "name":
"alternate"}, {"type": {"items": "CalledGenotype", "type": "array"},
"name": "calledGenotypes"}, {"doc": "", "type": {"items":
"ReportEvent", "type": "array"}, "name": "reportEvents"}, {"doc": "",
"type": ["null", "float"], "name": "internalScore"}, {"type": ["null",
"VariantClassification"], "name": "variantClassification"}, {"doc":
"", "type": ["null", {"values": "string", "type": "map"}], "name":
"additionalTextualVariantAnnotations"}, {"doc": "", "type": ["null",
{"values": "float", "type": "map"}], "name":
"additionalNumericVariantAnnotations"}, {"doc": "", "type": ["null",
{"items": "string", "type": "array"}], "name": "comments"}], "type":
"record", "name": "ReportedStructuralVariant"}, "type": "array"}],
"name": "reportedStructuralVariants"}, {"doc": "", "type": ["null",
{"items": "string", "type": "array"}], "name": "comments"}]}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "analysisId",
        "comments",
        "companyName",
        "reportRequestId",
        "reportURI",
        "reportedStructuralVariants",
        "reportedVariants",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'reportedVariants': ReportedVariant,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'reportedVariants': ReportedVariant,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'analysisId', 'comments', 'companyName', 'reportRequestId',
        'reportURI', 'reportedStructuralVariants', 'reportedVariants'
    ]

    def __init__(self, **kwargs):
        self.analysisId = kwargs.get(
            'analysisId', None)
        self.comments = kwargs.get(
            'comments', None)
        self.companyName = kwargs.get(
            'companyName', None)
        self.reportRequestId = kwargs.get(
            'reportRequestId', None)
        self.reportURI = kwargs.get(
            'reportURI', None)
        self.reportedStructuralVariants = kwargs.get(
            'reportedStructuralVariants', None)
        self.reportedVariants = kwargs.get(
            'reportedVariants', None)


class InterpretationRequest(ProtocolElement):
    """
    This record represents basic information for this report
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"InterpretationRequest", "fields": [{"doc": "", "type": "string",
"name": "reportRequestId"}, {"doc": "", "type": "int", "name":
"reportVersion"}, {"default": false, "doc": "", "type": "boolean",
"name": "interpretGenome"}, {"doc": "", "type": {"items": {"doc": "",
"type": "record", "name": "File", "fields": [{"doc": "", "type":
["null", "string", {"items": "string", "type": "array"}], "name":
"SampleId"}, {"doc": "", "type": "string", "name": "URIFile"}]},
"type": "array"}, "name": "BAMs"}, {"doc": "", "type": {"items":
"File", "type": "array"}, "name": "gVCFs"}, {"doc": "", "type":
{"items": "File", "type": "array"}, "name": "VCFs"}, {"doc": "",
"type": {"items": "File", "type": "array"}, "name": "BigWigs"},
{"doc": "", "type": "File", "name": "annotationFile"}, {"type":
{"fields": [{"doc": "", "type": "string", "name":
"recruitedDiseaseCategory"}, {"doc": "", "type": "string", "name":
"dataModelCatalogueVersion"}, {"type": {"doc": "", "type": "record",
"name": "Pedigree", "fields": [{"doc": "", "type": "string", "name":
"familyId"}, {"type": {"items": {"doc": "", "type": "record", "name":
"RDParticipant", "fields": [{"type": "string", "name": "id"}, {"doc":
"", "type": {"symbols": ["male", "female", "unknown"], "type": "enum",
"name": "Sex"}, "name": "sex"}, {"doc": "", "type": ["null",
"string"], "name": "father"}, {"type": ["null", "string"], "name":
"mother"}, {"type": "string", "name": "familyId"}, {"type": ["null",
{"items": "string", "type": "array"}], "name": "externalIds"}, {"doc":
"", "type": ["null", "int"], "name": "twinGroup"}, {"doc": "", "type":
["null", "boolean"], "name": "monozygotic"}, {"type": ["null",
{"symbols": ["not_adopted", "adoptedin", "adoptedout"], "doc": "",
"type": "enum", "name": "AdoptedStatus"}], "name": "adoptedStatus"},
{"doc": "", "type": ["null", {"symbols": ["aborted", "deceased",
"unborn", "stillborn", "miscarriage"], "type": "enum", "name":
"LifeStatus"}], "name": "lifeStatus"}, {"doc": "", "type": ["null",
"boolean"], "name": "consanguineousParents"}, {"doc": "", "type":
["null", "boolean"], "name": "consanguineousPopulation"}, {"doc": "",
"type": ["null", {"symbols": ["unaffected", "affected", "carrier",
"uncertain", "presymptomatic"], "doc": "", "type": "enum", "name":
"CarrierStatus"}], "name": "carrierStatus"}, {"doc": "", "type":
{"items": {"doc": "", "type": "record", "name": "Disorder", "fields":
[{"doc": "", "type": ["null", "string"], "name": "diseaseGroup"},
{"doc": "", "type": ["null", "string"], "name": "diseaseSubGroup"},
{"doc": "", "type": ["null", "string"], "name": "specificDisease"},
{"doc": "", "type": ["null", "int"], "name": "ageOfOnset"}]}, "type":
"array"}, "name": "disorderList"}, {"doc": "", "type": {"items":
{"doc": "", "type": "record", "name": "HpoTerm", "fields": [{"doc":
"", "type": "string", "name": "term"}, {"doc": "", "type": ["null",
"boolean"], "name": "termPresence"}, {"doc": "", "type": ["null",
{"items": "string", "type": "array"}], "name": "modifier"}, {"doc":
"", "type": ["null", "int"], "name": "ageOfOnset"}]}, "type":
"array"}, "name": "hpoTermList"}, {"default": "v4.2", "doc": "",
"type": "string", "name": "dataModelCatalogueVersion"}, {"doc": "",
"type": {"fields": [{"doc": "", "type": {"symbols": ["pilot", "main"],
"doc": "", "type": "enum", "name": "ProgramPhase"}, "name":
"programPhase"}, {"default": false, "doc": "", "type": "boolean",
"name": "optOut"}, {"type": ["null", "string"], "name":
"consentVersion"}, {"default": false, "type": "boolean", "name":
"programmeConsent"}, {"default": false, "doc": "", "type": "boolean",
"name": "primaryFindingConsent"}, {"default": false, "doc": "",
"type": "boolean", "name": "secondaryFindingConsent"}, {"default":
false, "doc": "", "type": "boolean", "name": "carrierStatusConsent"}],
"type": "record", "name": "ConsentStatus"}, "name": "consentStatus"},
{"doc": "", "type": ["null", {"items": "string", "type": "array"}],
"name": "samples"}, {"doc": "", "type": ["null", {"values": "string",
"type": "map"}], "name": "additionalInformation"}]}, "type": "array"},
"name": "participants"}]}, "name": "pedigree"}, {"doc": "", "type":
["null", "File"], "name": "inbreedingCoefficientEstimates"}, {"doc":
"", "type": ["null", {"items": "string", "type": "array"}], "name":
"ancestries"}, {"doc": "", "type": ["null", "string"], "name":
"complexGeneticPhenomena"}, {"doc": "", "type": {"items": "string",
"type": "array"}, "name": "modeOfInheritance"}, {"doc": "", "type":
["null", {"doc": "", "type": "record", "name": "VirtualPanel",
"fields": [{"type": "string", "name": "SpecificDiseaseTitle"}, {"doc":
"", "type": ["null", {"items": "string", "type": "array"}], "name":
"geneIds"}, {"doc": "", "type": ["null", {"items": "string", "type":
"array"}], "name": "clinicalRelevantTranscripts"}, {"doc": "", "type":
["null", "File"], "name": "relevantRegions"}, {"doc": "", "type":
["null", "File"], "name": "clinicalRelevantVariants"}]}], "name":
"virtualPanel"}, {"doc": "", "type": ["null", {"doc": "", "type":
"record", "name": "OtherFamilyHistory", "fields": [{"type": ["null",
{"items": "string", "type": "array"}], "name":
"maternalFamilyHistory"}, {"type": ["null", {"items": "string",
"type": "array"}], "name": "paternalFamilyHistory"}]}], "name":
"otherFamilyHistory"}], "type": "record", "name": "RareDisease"},
"name": "rareDisease"}, {"type": "string", "name": "analysisURI"},
{"type": "string", "name": "analysisVersion"}, {"doc": "", "type":
["null", {"values": "string", "type": "map"}], "name":
"additionalInfo"}], "doc": ""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "BAMs",
        "BigWigs",
        "VCFs",
        "additionalInfo",
        "analysisURI",
        "analysisVersion",
        "annotationFile",
        "gVCFs",
        "rareDisease",
        "reportRequestId",
        "reportVersion",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'BAMs': File,
            'BigWigs': File,
            'VCFs': File,
            'annotationFile': File,
            'gVCFs': File,
            'rareDisease': RareDisease,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'BAMs': File,
            'BigWigs': File,
            'VCFs': File,
            'annotationFile': File,
            'gVCFs': File,
            'rareDisease': RareDisease,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'BAMs', 'BigWigs', 'VCFs', 'additionalInfo', 'analysisURI',
        'analysisVersion', 'annotationFile', 'gVCFs',
        'interpretGenome', 'rareDisease', 'reportRequestId',
        'reportVersion'
    ]

    def __init__(self, **kwargs):
        self.BAMs = kwargs.get(
            'BAMs', None)
        self.BigWigs = kwargs.get(
            'BigWigs', None)
        self.VCFs = kwargs.get(
            'VCFs', None)
        self.additionalInfo = kwargs.get(
            'additionalInfo', None)
        self.analysisURI = kwargs.get(
            'analysisURI', None)
        self.analysisVersion = kwargs.get(
            'analysisVersion', None)
        self.annotationFile = kwargs.get(
            'annotationFile', None)
        self.gVCFs = kwargs.get(
            'gVCFs', None)
        self.interpretGenome = kwargs.get(
            'interpretGenome', False)
        self.rareDisease = kwargs.get(
            'rareDisease', None)
        self.reportRequestId = kwargs.get(
            'reportRequestId', None)
        self.reportVersion = kwargs.get(
            'reportVersion', None)


class LifeStatus(object):
    """
    No documentation
    """
    aborted = "aborted"
    deceased = "deceased"
    unborn = "unborn"
    stillborn = "stillborn"
    miscarriage = "miscarriage"


class MatchedSamples(ProtocolElement):
    """
    This define a pair of germline and tumor, this pair should/must be
    analyzed together
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"MatchedSamples", "fields": [{"type": "string", "name":
"germlineSampleId"}, {"type": "string", "name": "tumorSampleId"}],
"doc": ""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "germlineSampleId",
        "tumorSampleId",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'germlineSampleId', 'tumorSampleId'
    ]

    def __init__(self, **kwargs):
        self.germlineSampleId = kwargs.get(
            'germlineSampleId', None)
        self.tumorSampleId = kwargs.get(
            'tumorSampleId', None)


class Method(object):
    """
    TODO: perhaps we should change phase to stage? since we may wish
    to use Phase for Haplotype Phase?
    """
    resection = "resection"
    biopsy = "biopsy"


class OtherFamilyHistory(ProtocolElement):
    """
    Family history for secondary findings TODO: define the
    enumerations + other. Currently from model catalogue
    EndocrineTumours  colorectal  BreastOvarian  HDOrStroke Arrays of
    strings describing discrete family history phenotypes
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"OtherFamilyHistory", "fields": [{"type": ["null", {"items": "string",
"type": "array"}], "name": "maternalFamilyHistory"}, {"type": ["null",
{"items": "string", "type": "array"}], "name":
"paternalFamilyHistory"}], "doc": ""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "maternalFamilyHistory",
        "paternalFamilyHistory",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'maternalFamilyHistory', 'paternalFamilyHistory'
    ]

    def __init__(self, **kwargs):
        self.maternalFamilyHistory = kwargs.get(
            'maternalFamilyHistory', None)
        self.paternalFamilyHistory = kwargs.get(
            'paternalFamilyHistory', None)


class Pedigree(ProtocolElement):
    """
    This is the concept of a family with associated phenotypes as
    present in the record RDParticipant
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"Pedigree", "fields": [{"doc": "", "type": "string", "name":
"familyId"}, {"type": {"items": {"doc": "", "type": "record", "name":
"RDParticipant", "fields": [{"type": "string", "name": "id"}, {"doc":
"", "type": {"symbols": ["male", "female", "unknown"], "type": "enum",
"name": "Sex"}, "name": "sex"}, {"doc": "", "type": ["null",
"string"], "name": "father"}, {"type": ["null", "string"], "name":
"mother"}, {"type": "string", "name": "familyId"}, {"type": ["null",
{"items": "string", "type": "array"}], "name": "externalIds"}, {"doc":
"", "type": ["null", "int"], "name": "twinGroup"}, {"doc": "", "type":
["null", "boolean"], "name": "monozygotic"}, {"type": ["null",
{"symbols": ["not_adopted", "adoptedin", "adoptedout"], "doc": "",
"type": "enum", "name": "AdoptedStatus"}], "name": "adoptedStatus"},
{"doc": "", "type": ["null", {"symbols": ["aborted", "deceased",
"unborn", "stillborn", "miscarriage"], "type": "enum", "name":
"LifeStatus"}], "name": "lifeStatus"}, {"doc": "", "type": ["null",
"boolean"], "name": "consanguineousParents"}, {"doc": "", "type":
["null", "boolean"], "name": "consanguineousPopulation"}, {"doc": "",
"type": ["null", {"symbols": ["unaffected", "affected", "carrier",
"uncertain", "presymptomatic"], "doc": "", "type": "enum", "name":
"CarrierStatus"}], "name": "carrierStatus"}, {"doc": "", "type":
{"items": {"doc": "", "type": "record", "name": "Disorder", "fields":
[{"doc": "", "type": ["null", "string"], "name": "diseaseGroup"},
{"doc": "", "type": ["null", "string"], "name": "diseaseSubGroup"},
{"doc": "", "type": ["null", "string"], "name": "specificDisease"},
{"doc": "", "type": ["null", "int"], "name": "ageOfOnset"}]}, "type":
"array"}, "name": "disorderList"}, {"doc": "", "type": {"items":
{"doc": "", "type": "record", "name": "HpoTerm", "fields": [{"doc":
"", "type": "string", "name": "term"}, {"doc": "", "type": ["null",
"boolean"], "name": "termPresence"}, {"doc": "", "type": ["null",
{"items": "string", "type": "array"}], "name": "modifier"}, {"doc":
"", "type": ["null", "int"], "name": "ageOfOnset"}]}, "type":
"array"}, "name": "hpoTermList"}, {"default": "v4.2", "doc": "",
"type": "string", "name": "dataModelCatalogueVersion"}, {"doc": "",
"type": {"fields": [{"doc": "", "type": {"symbols": ["pilot", "main"],
"doc": "", "type": "enum", "name": "ProgramPhase"}, "name":
"programPhase"}, {"default": false, "doc": "", "type": "boolean",
"name": "optOut"}, {"type": ["null", "string"], "name":
"consentVersion"}, {"default": false, "type": "boolean", "name":
"programmeConsent"}, {"default": false, "doc": "", "type": "boolean",
"name": "primaryFindingConsent"}, {"default": false, "doc": "",
"type": "boolean", "name": "secondaryFindingConsent"}, {"default":
false, "doc": "", "type": "boolean", "name": "carrierStatusConsent"}],
"type": "record", "name": "ConsentStatus"}, "name": "consentStatus"},
{"doc": "", "type": ["null", {"items": "string", "type": "array"}],
"name": "samples"}, {"doc": "", "type": ["null", {"values": "string",
"type": "map"}], "name": "additionalInformation"}]}, "type": "array"},
"name": "participants"}], "doc": ""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "familyId",
        "participants",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'participants': RDParticipant,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'participants': RDParticipant,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'familyId', 'participants'
    ]

    def __init__(self, **kwargs):
        self.familyId = kwargs.get(
            'familyId', None)
        self.participants = kwargs.get(
            'participants', None)


class Penetrance(object):
    """
    Penetrance assumed in the analysis
    """
    complete = "complete"
    incomplete = "incomplete"


class Phase(object):
    """
    TODO: check in remission....
    """
    primary = "primary"
    metastasis = "metastasis"


class PreservationMethod(object):
    """
    No documentation
    """
    FFPE = "FFPE"
    FF = "FF"
    Unknown = "Unknown"
    LEUK = "LEUK"
    GL = "GL"


class ProgramPhase(object):
    """
    This is to distinguish into which part of the program the
    participant was consented to
    """
    pilot = "pilot"
    main = "main"


class RDParticipant(ProtocolElement):
    """
    This defines a RD Participant (demographics and pedigree
    information)
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"RDParticipant", "fields": [{"type": "string", "name": "id"}, {"doc":
"", "type": {"symbols": ["male", "female", "unknown"], "type": "enum",
"name": "Sex"}, "name": "sex"}, {"doc": "", "type": ["null",
"string"], "name": "father"}, {"type": ["null", "string"], "name":
"mother"}, {"type": "string", "name": "familyId"}, {"type": ["null",
{"items": "string", "type": "array"}], "name": "externalIds"}, {"doc":
"", "type": ["null", "int"], "name": "twinGroup"}, {"doc": "", "type":
["null", "boolean"], "name": "monozygotic"}, {"type": ["null",
{"symbols": ["not_adopted", "adoptedin", "adoptedout"], "doc": "",
"type": "enum", "name": "AdoptedStatus"}], "name": "adoptedStatus"},
{"doc": "", "type": ["null", {"symbols": ["aborted", "deceased",
"unborn", "stillborn", "miscarriage"], "type": "enum", "name":
"LifeStatus"}], "name": "lifeStatus"}, {"doc": "", "type": ["null",
"boolean"], "name": "consanguineousParents"}, {"doc": "", "type":
["null", "boolean"], "name": "consanguineousPopulation"}, {"doc": "",
"type": ["null", {"symbols": ["unaffected", "affected", "carrier",
"uncertain", "presymptomatic"], "doc": "", "type": "enum", "name":
"CarrierStatus"}], "name": "carrierStatus"}, {"doc": "", "type":
{"items": {"doc": "", "type": "record", "name": "Disorder", "fields":
[{"doc": "", "type": ["null", "string"], "name": "diseaseGroup"},
{"doc": "", "type": ["null", "string"], "name": "diseaseSubGroup"},
{"doc": "", "type": ["null", "string"], "name": "specificDisease"},
{"doc": "", "type": ["null", "int"], "name": "ageOfOnset"}]}, "type":
"array"}, "name": "disorderList"}, {"doc": "", "type": {"items":
{"doc": "", "type": "record", "name": "HpoTerm", "fields": [{"doc":
"", "type": "string", "name": "term"}, {"doc": "", "type": ["null",
"boolean"], "name": "termPresence"}, {"doc": "", "type": ["null",
{"items": "string", "type": "array"}], "name": "modifier"}, {"doc":
"", "type": ["null", "int"], "name": "ageOfOnset"}]}, "type":
"array"}, "name": "hpoTermList"}, {"default": "v4.2", "doc": "",
"type": "string", "name": "dataModelCatalogueVersion"}, {"doc": "",
"type": {"fields": [{"doc": "", "type": {"symbols": ["pilot", "main"],
"doc": "", "type": "enum", "name": "ProgramPhase"}, "name":
"programPhase"}, {"default": false, "doc": "", "type": "boolean",
"name": "optOut"}, {"type": ["null", "string"], "name":
"consentVersion"}, {"default": false, "type": "boolean", "name":
"programmeConsent"}, {"default": false, "doc": "", "type": "boolean",
"name": "primaryFindingConsent"}, {"default": false, "doc": "",
"type": "boolean", "name": "secondaryFindingConsent"}, {"default":
false, "doc": "", "type": "boolean", "name": "carrierStatusConsent"}],
"type": "record", "name": "ConsentStatus"}, "name": "consentStatus"},
{"doc": "", "type": ["null", {"items": "string", "type": "array"}],
"name": "samples"}, {"doc": "", "type": ["null", {"values": "string",
"type": "map"}], "name": "additionalInformation"}], "doc": ""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "additionalInformation",
        "adoptedStatus",
        "carrierStatus",
        "consanguineousParents",
        "consanguineousPopulation",
        "consentStatus",
        "disorderList",
        "externalIds",
        "familyId",
        "father",
        "hpoTermList",
        "id",
        "lifeStatus",
        "monozygotic",
        "mother",
        "samples",
        "sex",
        "twinGroup",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'consentStatus': ConsentStatus,
            'disorderList': Disorder,
            'hpoTermList': HpoTerm,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'consentStatus': ConsentStatus,
            'disorderList': Disorder,
            'hpoTermList': HpoTerm,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'additionalInformation', 'adoptedStatus', 'carrierStatus',
        'consanguineousParents', 'consanguineousPopulation',
        'consentStatus', 'dataModelCatalogueVersion', 'disorderList',
        'externalIds', 'familyId', 'father', 'hpoTermList', 'id',
        'lifeStatus', 'monozygotic', 'mother', 'samples', 'sex',
        'twinGroup'
    ]

    def __init__(self, **kwargs):
        self.additionalInformation = kwargs.get(
            'additionalInformation', None)
        self.adoptedStatus = kwargs.get(
            'adoptedStatus', None)
        self.carrierStatus = kwargs.get(
            'carrierStatus', None)
        self.consanguineousParents = kwargs.get(
            'consanguineousParents', None)
        self.consanguineousPopulation = kwargs.get(
            'consanguineousPopulation', None)
        self.consentStatus = kwargs.get(
            'consentStatus', None)
        self.dataModelCatalogueVersion = kwargs.get(
            'dataModelCatalogueVersion', v4.2)
        self.disorderList = kwargs.get(
            'disorderList', None)
        self.externalIds = kwargs.get(
            'externalIds', None)
        self.familyId = kwargs.get(
            'familyId', None)
        self.father = kwargs.get(
            'father', None)
        self.hpoTermList = kwargs.get(
            'hpoTermList', None)
        self.id = kwargs.get(
            'id', None)
        self.lifeStatus = kwargs.get(
            'lifeStatus', None)
        self.monozygotic = kwargs.get(
            'monozygotic', None)
        self.mother = kwargs.get(
            'mother', None)
        self.samples = kwargs.get(
            'samples', None)
        self.sex = kwargs.get(
            'sex', None)
        self.twinGroup = kwargs.get(
            'twinGroup', None)


class RareDisease(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"RareDisease", "fields": [{"doc": "", "type": "string", "name":
"recruitedDiseaseCategory"}, {"doc": "", "type": "string", "name":
"dataModelCatalogueVersion"}, {"type": {"doc": "", "type": "record",
"name": "Pedigree", "fields": [{"doc": "", "type": "string", "name":
"familyId"}, {"type": {"items": {"doc": "", "type": "record", "name":
"RDParticipant", "fields": [{"type": "string", "name": "id"}, {"doc":
"", "type": {"symbols": ["male", "female", "unknown"], "type": "enum",
"name": "Sex"}, "name": "sex"}, {"doc": "", "type": ["null",
"string"], "name": "father"}, {"type": ["null", "string"], "name":
"mother"}, {"type": "string", "name": "familyId"}, {"type": ["null",
{"items": "string", "type": "array"}], "name": "externalIds"}, {"doc":
"", "type": ["null", "int"], "name": "twinGroup"}, {"doc": "", "type":
["null", "boolean"], "name": "monozygotic"}, {"type": ["null",
{"symbols": ["not_adopted", "adoptedin", "adoptedout"], "doc": "",
"type": "enum", "name": "AdoptedStatus"}], "name": "adoptedStatus"},
{"doc": "", "type": ["null", {"symbols": ["aborted", "deceased",
"unborn", "stillborn", "miscarriage"], "type": "enum", "name":
"LifeStatus"}], "name": "lifeStatus"}, {"doc": "", "type": ["null",
"boolean"], "name": "consanguineousParents"}, {"doc": "", "type":
["null", "boolean"], "name": "consanguineousPopulation"}, {"doc": "",
"type": ["null", {"symbols": ["unaffected", "affected", "carrier",
"uncertain", "presymptomatic"], "doc": "", "type": "enum", "name":
"CarrierStatus"}], "name": "carrierStatus"}, {"doc": "", "type":
{"items": {"doc": "", "type": "record", "name": "Disorder", "fields":
[{"doc": "", "type": ["null", "string"], "name": "diseaseGroup"},
{"doc": "", "type": ["null", "string"], "name": "diseaseSubGroup"},
{"doc": "", "type": ["null", "string"], "name": "specificDisease"},
{"doc": "", "type": ["null", "int"], "name": "ageOfOnset"}]}, "type":
"array"}, "name": "disorderList"}, {"doc": "", "type": {"items":
{"doc": "", "type": "record", "name": "HpoTerm", "fields": [{"doc":
"", "type": "string", "name": "term"}, {"doc": "", "type": ["null",
"boolean"], "name": "termPresence"}, {"doc": "", "type": ["null",
{"items": "string", "type": "array"}], "name": "modifier"}, {"doc":
"", "type": ["null", "int"], "name": "ageOfOnset"}]}, "type":
"array"}, "name": "hpoTermList"}, {"default": "v4.2", "doc": "",
"type": "string", "name": "dataModelCatalogueVersion"}, {"doc": "",
"type": {"fields": [{"doc": "", "type": {"symbols": ["pilot", "main"],
"doc": "", "type": "enum", "name": "ProgramPhase"}, "name":
"programPhase"}, {"default": false, "doc": "", "type": "boolean",
"name": "optOut"}, {"type": ["null", "string"], "name":
"consentVersion"}, {"default": false, "type": "boolean", "name":
"programmeConsent"}, {"default": false, "doc": "", "type": "boolean",
"name": "primaryFindingConsent"}, {"default": false, "doc": "",
"type": "boolean", "name": "secondaryFindingConsent"}, {"default":
false, "doc": "", "type": "boolean", "name": "carrierStatusConsent"}],
"type": "record", "name": "ConsentStatus"}, "name": "consentStatus"},
{"doc": "", "type": ["null", {"items": "string", "type": "array"}],
"name": "samples"}, {"doc": "", "type": ["null", {"values": "string",
"type": "map"}], "name": "additionalInformation"}]}, "type": "array"},
"name": "participants"}]}, "name": "pedigree"}, {"doc": "", "type":
["null", {"doc": "", "type": "record", "name": "File", "fields":
[{"doc": "", "type": ["null", "string", {"items": "string", "type":
"array"}], "name": "SampleId"}, {"doc": "", "type": "string", "name":
"URIFile"}]}], "name": "inbreedingCoefficientEstimates"}, {"doc": "",
"type": ["null", {"items": "string", "type": "array"}], "name":
"ancestries"}, {"doc": "", "type": ["null", "string"], "name":
"complexGeneticPhenomena"}, {"doc": "", "type": {"items": "string",
"type": "array"}, "name": "modeOfInheritance"}, {"doc": "", "type":
["null", {"doc": "", "type": "record", "name": "VirtualPanel",
"fields": [{"type": "string", "name": "SpecificDiseaseTitle"}, {"doc":
"", "type": ["null", {"items": "string", "type": "array"}], "name":
"geneIds"}, {"doc": "", "type": ["null", {"items": "string", "type":
"array"}], "name": "clinicalRelevantTranscripts"}, {"doc": "", "type":
["null", "File"], "name": "relevantRegions"}, {"doc": "", "type":
["null", "File"], "name": "clinicalRelevantVariants"}]}], "name":
"virtualPanel"}, {"doc": "", "type": ["null", {"doc": "", "type":
"record", "name": "OtherFamilyHistory", "fields": [{"type": ["null",
{"items": "string", "type": "array"}], "name":
"maternalFamilyHistory"}, {"type": ["null", {"items": "string",
"type": "array"}], "name": "paternalFamilyHistory"}]}], "name":
"otherFamilyHistory"}]}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "ancestries",
        "complexGeneticPhenomena",
        "dataModelCatalogueVersion",
        "inbreedingCoefficientEstimates",
        "modeOfInheritance",
        "otherFamilyHistory",
        "pedigree",
        "recruitedDiseaseCategory",
        "virtualPanel",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'inbreedingCoefficientEstimates': File,
            'otherFamilyHistory': OtherFamilyHistory,
            'pedigree': Pedigree,
            'virtualPanel': VirtualPanel,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'inbreedingCoefficientEstimates': File,
            'otherFamilyHistory': OtherFamilyHistory,
            'pedigree': Pedigree,
            'virtualPanel': VirtualPanel,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'ancestries', 'complexGeneticPhenomena',
        'dataModelCatalogueVersion', 'inbreedingCoefficientEstimates',
        'modeOfInheritance', 'otherFamilyHistory', 'pedigree',
        'recruitedDiseaseCategory', 'virtualPanel'
    ]

    def __init__(self, **kwargs):
        self.ancestries = kwargs.get(
            'ancestries', None)
        self.complexGeneticPhenomena = kwargs.get(
            'complexGeneticPhenomena', None)
        self.dataModelCatalogueVersion = kwargs.get(
            'dataModelCatalogueVersion', None)
        self.inbreedingCoefficientEstimates = kwargs.get(
            'inbreedingCoefficientEstimates', None)
        self.modeOfInheritance = kwargs.get(
            'modeOfInheritance', None)
        self.otherFamilyHistory = kwargs.get(
            'otherFamilyHistory', None)
        self.pedigree = kwargs.get(
            'pedigree', None)
        self.recruitedDiseaseCategory = kwargs.get(
            'recruitedDiseaseCategory', None)
        self.virtualPanel = kwargs.get(
            'virtualPanel', None)


class ReportEvent(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"ReportEvent", "fields": [{"doc": "", "type": "string", "name":
"modeOfInheritance"}, {"doc": "", "type": {"fields": [{"doc": "",
"type": "string", "name": "symbol"}, {"doc": "", "type": "string",
"name": "transcriptId"}, {"doc": "", "type": {"values": "string",
"type": "map"}, "name": "ids"}, {"doc": "", "type": ["null", "int"],
"name": "exomeSize"}, {"doc": "", "type": ["null", "float"], "name":
"geneCoverage"}], "type": "record", "name": "Gene"}, "name": "gene"},
{"doc": "", "type": {"symbols": ["complete", "incomplete"], "doc": "",
"type": "enum", "name": "Penetrance"}, "name": "penetrance"}, {"doc":
"", "type": "float", "name": "score"}, {"doc": "", "type": ["null",
"boolean"], "name": "fullyExplainsPhenotype"}, {"doc": "", "type":
["null", "int"], "name": "groupOfVariants"}, {"doc": "", "type":
"string", "name": "variantJustification"}, {"type": ["null",
{"symbols": ["NONE", "TIER1", "TIER2", "TIER3"], "doc": "", "type":
"enum", "name": "Tier"}], "name": "tier"}]}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "fullyExplainsPhenotype",
        "gene",
        "groupOfVariants",
        "modeOfInheritance",
        "penetrance",
        "score",
        "tier",
        "variantJustification",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'gene': Gene,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'gene': Gene,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'fullyExplainsPhenotype', 'gene', 'groupOfVariants',
        'modeOfInheritance', 'penetrance', 'score', 'tier',
        'variantJustification'
    ]

    def __init__(self, **kwargs):
        self.fullyExplainsPhenotype = kwargs.get(
            'fullyExplainsPhenotype', None)
        self.gene = kwargs.get(
            'gene', None)
        self.groupOfVariants = kwargs.get(
            'groupOfVariants', None)
        self.modeOfInheritance = kwargs.get(
            'modeOfInheritance', None)
        self.penetrance = kwargs.get(
            'penetrance', None)
        self.score = kwargs.get(
            'score', None)
        self.tier = kwargs.get(
            'tier', None)
        self.variantJustification = kwargs.get(
            'variantJustification', None)


class ReportEvents(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"ReportEvents", "fields": [{"doc": "", "type": ["null", "string"],
"name": "modeOfInheritance"}, {"doc": "", "type": {"fields": [{"doc":
"", "type": "string", "name": "symbol"}, {"doc": "", "type": "string",
"name": "transcriptId"}, {"doc": "", "type": {"values": "string",
"type": "map"}, "name": "ids"}, {"doc": "", "type": ["null", "int"],
"name": "exomeSize"}, {"doc": "", "type": ["null", "double"], "name":
"geneCoverage"}], "type": "record", "name": "Gene"}, "name": "gene"},
{"doc": "", "type": "float", "name": "score"}, {"doc": "", "type":
["null", "boolean"], "name": "fullyExplainsPhenotype"}, {"doc": "",
"type": ["null", "int"], "name": "groupOfVariants"}, {"doc": "",
"type": "string", "name": "variantJustification"}, {"type": ["null",
{"symbols": ["NONE", "TIER1", "TIER2", "TIER3"], "doc": "", "type":
"enum", "name": "Tier"}], "name": "tier"}]}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "fullyExplainsPhenotype",
        "gene",
        "groupOfVariants",
        "modeOfInheritance",
        "score",
        "tier",
        "variantJustification",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'gene': Gene,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'gene': Gene,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'fullyExplainsPhenotype', 'gene', 'groupOfVariants',
        'modeOfInheritance', 'score', 'tier', 'variantJustification'
    ]

    def __init__(self, **kwargs):
        self.fullyExplainsPhenotype = kwargs.get(
            'fullyExplainsPhenotype', None)
        self.gene = kwargs.get(
            'gene', None)
        self.groupOfVariants = kwargs.get(
            'groupOfVariants', None)
        self.modeOfInheritance = kwargs.get(
            'modeOfInheritance', None)
        self.score = kwargs.get(
            'score', None)
        self.tier = kwargs.get(
            'tier', None)
        self.variantJustification = kwargs.get(
            'variantJustification', None)


class ReportType(object):
    """
    No documentation
    """
    primaryFindings = "primaryFindings"
    SecondaryFindings = "SecondaryFindings"
    CarrierStatus = "CarrierStatus"


class ReportedStructuralVariant(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"ReportedStructuralVariant", "fields": [{"doc": "", "type": "string",
"name": "chromosome"}, {"doc": "", "type": "int", "name": "start"},
{"doc": "", "type": "int", "name": "end"}, {"doc": "", "type":
"string", "name": "type"}, {"doc": "", "type": "string", "name":
"reference"}, {"doc": "", "type": "string", "name": "alternate"},
{"type": {"items": {"doc": "", "type": "record", "name":
"CalledGenotype", "fields": [{"doc": "", "type": "string", "name":
"participantId"}, {"doc": "", "type": "string", "name": "genotype"},
{"doc": "", "type": ["null", "float"], "name": "alleleFraction"},
{"doc": "", "type": ["null", "int"], "name": "copyNumber"}]}, "type":
"array"}, "name": "calledGenotypes"}, {"doc": "", "type": {"items":
{"fields": [{"doc": "", "type": "string", "name":
"modeOfInheritance"}, {"doc": "", "type": {"fields": [{"doc": "",
"type": "string", "name": "symbol"}, {"doc": "", "type": "string",
"name": "transcriptId"}, {"doc": "", "type": {"values": "string",
"type": "map"}, "name": "ids"}, {"doc": "", "type": ["null", "int"],
"name": "exomeSize"}, {"doc": "", "type": ["null", "float"], "name":
"geneCoverage"}], "type": "record", "name": "Gene"}, "name": "gene"},
{"doc": "", "type": {"symbols": ["complete", "incomplete"], "doc": "",
"type": "enum", "name": "Penetrance"}, "name": "penetrance"}, {"doc":
"", "type": "float", "name": "score"}, {"doc": "", "type": ["null",
"boolean"], "name": "fullyExplainsPhenotype"}, {"doc": "", "type":
["null", "int"], "name": "groupOfVariants"}, {"doc": "", "type":
"string", "name": "variantJustification"}, {"type": ["null",
{"symbols": ["NONE", "TIER1", "TIER2", "TIER3"], "doc": "", "type":
"enum", "name": "Tier"}], "name": "tier"}], "type": "record", "name":
"ReportEvent"}, "type": "array"}, "name": "reportEvents"}, {"doc": "",
"type": ["null", "float"], "name": "internalScore"}, {"type": ["null",
{"symbols": ["BENIGN", "LIKELY_BENIGN", "VUS", "LIKELY_PATHOGENIC",
"PATHOGENIC"], "doc": "", "type": "enum", "name":
"VariantClassification"}], "name": "variantClassification"}, {"doc":
"", "type": ["null", {"values": "string", "type": "map"}], "name":
"additionalTextualVariantAnnotations"}, {"doc": "", "type": ["null",
{"values": "float", "type": "map"}], "name":
"additionalNumericVariantAnnotations"}, {"doc": "", "type": ["null",
{"items": "string", "type": "array"}], "name": "comments"}]}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "additionalNumericVariantAnnotations",
        "additionalTextualVariantAnnotations",
        "alternate",
        "calledGenotypes",
        "chromosome",
        "comments",
        "end",
        "internalScore",
        "reference",
        "reportEvents",
        "start",
        "type",
        "variantClassification",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'calledGenotypes': CalledGenotype,
            'reportEvents': ReportEvent,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'calledGenotypes': CalledGenotype,
            'reportEvents': ReportEvent,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'additionalNumericVariantAnnotations',
        'additionalTextualVariantAnnotations', 'alternate',
        'calledGenotypes', 'chromosome', 'comments', 'end',
        'internalScore', 'reference', 'reportEvents', 'start', 'type',
        'variantClassification'
    ]

    def __init__(self, **kwargs):
        self.additionalNumericVariantAnnotations = kwargs.get(
            'additionalNumericVariantAnnotations', None)
        self.additionalTextualVariantAnnotations = kwargs.get(
            'additionalTextualVariantAnnotations', None)
        self.alternate = kwargs.get(
            'alternate', None)
        self.calledGenotypes = kwargs.get(
            'calledGenotypes', None)
        self.chromosome = kwargs.get(
            'chromosome', None)
        self.comments = kwargs.get(
            'comments', None)
        self.end = kwargs.get(
            'end', None)
        self.internalScore = kwargs.get(
            'internalScore', None)
        self.reference = kwargs.get(
            'reference', None)
        self.reportEvents = kwargs.get(
            'reportEvents', None)
        self.start = kwargs.get(
            'start', None)
        self.type = kwargs.get(
            'type', None)
        self.variantClassification = kwargs.get(
            'variantClassification', None)


class ReportedStructuralVariants(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"ReportedStructuralVariants", "fields": [{"doc": "", "type": "string",
"name": "chromosome"}, {"doc": "", "type": "int", "name": "start"},
{"doc": "", "type": "int", "name": "end"}, {"doc": "", "type":
"string", "name": "type"}, {"doc": "", "type": "string", "name":
"reference"}, {"doc": "", "type": "string", "name": "alternate"},
{"type": {"items": {"fields": [{"doc": "", "type": ["null", "string"],
"name": "modeOfInheritance"}, {"doc": "", "type": {"fields": [{"doc":
"", "type": "string", "name": "symbol"}, {"doc": "", "type": "string",
"name": "transcriptId"}, {"doc": "", "type": {"values": "string",
"type": "map"}, "name": "ids"}, {"doc": "", "type": ["null", "int"],
"name": "exomeSize"}, {"doc": "", "type": ["null", "double"], "name":
"geneCoverage"}], "type": "record", "name": "Gene"}, "name": "gene"},
{"doc": "", "type": "float", "name": "score"}, {"doc": "", "type":
["null", "boolean"], "name": "fullyExplainsPhenotype"}, {"doc": "",
"type": ["null", "int"], "name": "groupOfVariants"}, {"doc": "",
"type": "string", "name": "variantJustification"}, {"type": ["null",
{"symbols": ["NONE", "TIER1", "TIER2", "TIER3"], "doc": "", "type":
"enum", "name": "Tier"}], "name": "tier"}], "type": "record", "name":
"ReportEvents"}, "type": "array"}, "name": "ReportEvents"}, {"type":
{"items": {"doc": "", "type": "record", "name": "CalledGenotypes",
"fields": [{"doc": "", "type": "string", "name": "sampleId"}, {"doc":
"", "type": "string", "name": "genotype"}, {"type": ["null", "float"],
"name": "alleleFraction"}, {"doc": "", "type": ["null", "int"],
"name": "copyNumber"}]}, "type": "array"}, "name": "calledGenotypes"},
{"doc": "", "type": ["null", "float"], "name": "internalScore"},
{"type": ["null", {"symbols": ["BENIGN", "LIKELY_BENIGN", "VUS",
"LIKELY_PATHOGENIC", "PATHOGENIC"], "doc": "", "type": "enum", "name":
"VariantClassification"}], "name": "variantClassification"}, {"doc":
"", "type": ["null", {"values": "string", "type": "map"}], "name":
"additionalTextualVariantAnnotations"}, {"doc": "", "type": ["null",
{"values": "float", "type": "map"}], "name":
"additionalNumericVariantAnnotations"}, {"doc": "", "type": ["null",
{"items": "string", "type": "array"}], "name": "comments"}, {"doc":
"", "type": {"symbols": ["unknown", "somatic", "germline"], "doc": "",
"type": "enum", "name": "SomaticOrGermline"}, "name":
"somaticOrGermline"}]}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "ReportEvents",
        "additionalNumericVariantAnnotations",
        "additionalTextualVariantAnnotations",
        "alternate",
        "calledGenotypes",
        "chromosome",
        "comments",
        "end",
        "internalScore",
        "reference",
        "somaticOrGermline",
        "start",
        "type",
        "variantClassification",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'ReportEvents': ReportEvents,
            'calledGenotypes': CalledGenotypes,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'ReportEvents': ReportEvents,
            'calledGenotypes': CalledGenotypes,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'ReportEvents', 'additionalNumericVariantAnnotations',
        'additionalTextualVariantAnnotations', 'alternate',
        'calledGenotypes', 'chromosome', 'comments', 'end',
        'internalScore', 'reference', 'somaticOrGermline', 'start',
        'type', 'variantClassification'
    ]

    def __init__(self, **kwargs):
        self.ReportEvents = kwargs.get(
            'ReportEvents', None)
        self.additionalNumericVariantAnnotations = kwargs.get(
            'additionalNumericVariantAnnotations', None)
        self.additionalTextualVariantAnnotations = kwargs.get(
            'additionalTextualVariantAnnotations', None)
        self.alternate = kwargs.get(
            'alternate', None)
        self.calledGenotypes = kwargs.get(
            'calledGenotypes', None)
        self.chromosome = kwargs.get(
            'chromosome', None)
        self.comments = kwargs.get(
            'comments', None)
        self.end = kwargs.get(
            'end', None)
        self.internalScore = kwargs.get(
            'internalScore', None)
        self.reference = kwargs.get(
            'reference', None)
        self.somaticOrGermline = kwargs.get(
            'somaticOrGermline', None)
        self.start = kwargs.get(
            'start', None)
        self.type = kwargs.get(
            'type', None)
        self.variantClassification = kwargs.get(
            'variantClassification', None)


class ReportedVariant(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"ReportedVariant", "fields": [{"doc": "", "type": "string", "name":
"chromosome"}, {"doc": "", "type": "string", "name": "dbSNPid"},
{"doc": "", "type": "int", "name": "position"}, {"doc": "", "type":
"string", "name": "reference"}, {"doc": "", "type": "string", "name":
"alternate"}, {"doc": "", "type": {"items": {"doc": "", "type":
"record", "name": "CalledGenotype", "fields": [{"doc": "", "type":
"string", "name": "participantId"}, {"doc": "", "type": "string",
"name": "genotype"}, {"doc": "", "type": ["null", "float"], "name":
"alleleFraction"}, {"doc": "", "type": ["null", "int"], "name":
"copyNumber"}]}, "type": "array"}, "name": "calledGenotypes"}, {"doc":
"", "type": {"items": {"fields": [{"doc": "", "type": "string",
"name": "modeOfInheritance"}, {"doc": "", "type": {"fields": [{"doc":
"", "type": "string", "name": "symbol"}, {"doc": "", "type": "string",
"name": "transcriptId"}, {"doc": "", "type": {"values": "string",
"type": "map"}, "name": "ids"}, {"doc": "", "type": ["null", "int"],
"name": "exomeSize"}, {"doc": "", "type": ["null", "float"], "name":
"geneCoverage"}], "type": "record", "name": "Gene"}, "name": "gene"},
{"doc": "", "type": {"symbols": ["complete", "incomplete"], "doc": "",
"type": "enum", "name": "Penetrance"}, "name": "penetrance"}, {"doc":
"", "type": "float", "name": "score"}, {"doc": "", "type": ["null",
"boolean"], "name": "fullyExplainsPhenotype"}, {"doc": "", "type":
["null", "int"], "name": "groupOfVariants"}, {"doc": "", "type":
"string", "name": "variantJustification"}, {"type": ["null",
{"symbols": ["NONE", "TIER1", "TIER2", "TIER3"], "doc": "", "type":
"enum", "name": "Tier"}], "name": "tier"}], "type": "record", "name":
"ReportEvent"}, "type": "array"}, "name": "reportEvents"}, {"doc": "",
"type": ["null", "float"], "name": "internalScore"}, {"type": ["null",
{"symbols": ["BENIGN", "LIKELY_BENIGN", "VUS", "LIKELY_PATHOGENIC",
"PATHOGENIC"], "doc": "", "type": "enum", "name":
"VariantClassification"}], "name": "variantClassification"}, {"doc":
"", "type": ["null", {"values": "string", "type": "map"}], "name":
"additionalTextualVariantAnnotations"}, {"doc": "", "type": ["null",
{"values": "float", "type": "map"}], "name":
"additionalNumericVariantAnnotations"}, {"doc": "", "type": ["null",
{"items": "string", "type": "array"}], "name": "comments"}]}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "additionalNumericVariantAnnotations",
        "additionalTextualVariantAnnotations",
        "alternate",
        "calledGenotypes",
        "chromosome",
        "comments",
        "dbSNPid",
        "internalScore",
        "position",
        "reference",
        "reportEvents",
        "variantClassification",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'calledGenotypes': CalledGenotype,
            'reportEvents': ReportEvent,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'calledGenotypes': CalledGenotype,
            'reportEvents': ReportEvent,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'additionalNumericVariantAnnotations',
        'additionalTextualVariantAnnotations', 'alternate',
        'calledGenotypes', 'chromosome', 'comments', 'dbSNPid',
        'internalScore', 'position', 'reference', 'reportEvents',
        'variantClassification'
    ]

    def __init__(self, **kwargs):
        self.additionalNumericVariantAnnotations = kwargs.get(
            'additionalNumericVariantAnnotations', None)
        self.additionalTextualVariantAnnotations = kwargs.get(
            'additionalTextualVariantAnnotations', None)
        self.alternate = kwargs.get(
            'alternate', None)
        self.calledGenotypes = kwargs.get(
            'calledGenotypes', None)
        self.chromosome = kwargs.get(
            'chromosome', None)
        self.comments = kwargs.get(
            'comments', None)
        self.dbSNPid = kwargs.get(
            'dbSNPid', None)
        self.internalScore = kwargs.get(
            'internalScore', None)
        self.position = kwargs.get(
            'position', None)
        self.reference = kwargs.get(
            'reference', None)
        self.reportEvents = kwargs.get(
            'reportEvents', None)
        self.variantClassification = kwargs.get(
            'variantClassification', None)


class ReportedVariants(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"ReportedVariants", "fields": [{"doc": "", "type": "string", "name":
"chromosome"}, {"doc": "", "type": "string", "name": "dbSNPid"},
{"doc": "", "type": "int", "name": "position"}, {"doc": "", "type":
"string", "name": "reference"}, {"doc": "", "type": "string", "name":
"alternate"}, {"doc": "", "type": {"items": {"fields": [{"doc": "",
"type": ["null", "string"], "name": "modeOfInheritance"}, {"doc": "",
"type": {"fields": [{"doc": "", "type": "string", "name": "symbol"},
{"doc": "", "type": "string", "name": "transcriptId"}, {"doc": "",
"type": {"values": "string", "type": "map"}, "name": "ids"}, {"doc":
"", "type": ["null", "int"], "name": "exomeSize"}, {"doc": "", "type":
["null", "double"], "name": "geneCoverage"}], "type": "record",
"name": "Gene"}, "name": "gene"}, {"doc": "", "type": "float", "name":
"score"}, {"doc": "", "type": ["null", "boolean"], "name":
"fullyExplainsPhenotype"}, {"doc": "", "type": ["null", "int"],
"name": "groupOfVariants"}, {"doc": "", "type": "string", "name":
"variantJustification"}, {"type": ["null", {"symbols": ["NONE",
"TIER1", "TIER2", "TIER3"], "doc": "", "type": "enum", "name":
"Tier"}], "name": "tier"}], "type": "record", "name": "ReportEvents"},
"type": "array"}, "name": "reportEvents"}, {"type": {"items": {"doc":
"", "type": "record", "name": "CalledGenotypes", "fields": [{"doc":
"", "type": "string", "name": "sampleId"}, {"doc": "", "type":
"string", "name": "genotype"}, {"type": ["null", "float"], "name":
"alleleFraction"}, {"doc": "", "type": ["null", "int"], "name":
"copyNumber"}]}, "type": "array"}, "name": "calledGenotypes"},
{"type": ["null", "float"], "name": "internalScore"}, {"type":
["null", {"symbols": ["BENIGN", "LIKELY_BENIGN", "VUS",
"LIKELY_PATHOGENIC", "PATHOGENIC"], "doc": "", "type": "enum", "name":
"VariantClassification"}], "name": "variantClassification"}, {"doc":
"", "type": ["null", {"values": "string", "type": "map"}], "name":
"additionalTextualVariantAnnotations"}, {"doc": "", "type": ["null",
{"values": "float", "type": "map"}], "name":
"additionalNumericVariantAnnotations"}, {"doc": "", "type": ["null",
{"items": "string", "type": "array"}], "name": "comments"}, {"doc":
"", "type": {"symbols": ["unknown", "somatic", "germline"], "doc": "",
"type": "enum", "name": "SomaticOrGermline"}, "name":
"somaticOrGermline"}]}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "additionalNumericVariantAnnotations",
        "additionalTextualVariantAnnotations",
        "alternate",
        "calledGenotypes",
        "chromosome",
        "comments",
        "dbSNPid",
        "internalScore",
        "position",
        "reference",
        "reportEvents",
        "somaticOrGermline",
        "variantClassification",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'calledGenotypes': CalledGenotypes,
            'reportEvents': ReportEvents,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'calledGenotypes': CalledGenotypes,
            'reportEvents': ReportEvents,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'additionalNumericVariantAnnotations',
        'additionalTextualVariantAnnotations', 'alternate',
        'calledGenotypes', 'chromosome', 'comments', 'dbSNPid',
        'internalScore', 'position', 'reference', 'reportEvents',
        'somaticOrGermline', 'variantClassification'
    ]

    def __init__(self, **kwargs):
        self.additionalNumericVariantAnnotations = kwargs.get(
            'additionalNumericVariantAnnotations', None)
        self.additionalTextualVariantAnnotations = kwargs.get(
            'additionalTextualVariantAnnotations', None)
        self.alternate = kwargs.get(
            'alternate', None)
        self.calledGenotypes = kwargs.get(
            'calledGenotypes', None)
        self.chromosome = kwargs.get(
            'chromosome', None)
        self.comments = kwargs.get(
            'comments', None)
        self.dbSNPid = kwargs.get(
            'dbSNPid', None)
        self.internalScore = kwargs.get(
            'internalScore', None)
        self.position = kwargs.get(
            'position', None)
        self.reference = kwargs.get(
            'reference', None)
        self.reportEvents = kwargs.get(
            'reportEvents', None)
        self.somaticOrGermline = kwargs.get(
            'somaticOrGermline', None)
        self.variantClassification = kwargs.get(
            'variantClassification', None)


class SampleType(object):
    """
    No documentation
    """
    germline = "germline"
    tumor = "tumor"


class Sex(object):
    """
    No documentation
    """
    male = "male"
    female = "female"
    unknown = "unknown"


class SomaticOrGermline(object):
    """
    this defines whether the variant reported is considered to be a
    somatic or germline finding.
    """
    unknown = "unknown"
    somatic = "somatic"
    germline = "germline"


class Tier(object):
    """
    Possible tiers as defined by Genomics England/ or Impact  - it is
    unlikely that tiers will be used for cancer, since these are gene
    specific, however impact classes as used to define the tiers may
    be
    """
    NONE = "NONE"
    TIER1 = "TIER1"
    TIER2 = "TIER2"
    TIER3 = "TIER3"


class VariantClassification(object):
    """
    This is the classification of the variant according to standard
    practice guidelines (e.g. ACMG)
    """
    BENIGN = "BENIGN"
    LIKELY_BENIGN = "LIKELY_BENIGN"
    VUS = "VUS"
    LIKELY_PATHOGENIC = "LIKELY_PATHOGENIC"
    PATHOGENIC = "PATHOGENIC"


class VirtualPanel(ProtocolElement):
    """
    This record defines a gene panel. The Level 4 Title is the name of
    the condition, for example Brugada syndrome
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"VirtualPanel", "fields": [{"type": "string", "name":
"SpecificDiseaseTitle"}, {"doc": "", "type": ["null", {"items":
"string", "type": "array"}], "name": "geneIds"}, {"doc": "", "type":
["null", {"items": "string", "type": "array"}], "name":
"clinicalRelevantTranscripts"}, {"doc": "", "type": ["null", {"doc":
"", "type": "record", "name": "File", "fields": [{"doc": "", "type":
["null", "string", {"items": "string", "type": "array"}], "name":
"SampleId"}, {"doc": "", "type": "string", "name": "URIFile"}]}],
"name": "relevantRegions"}, {"doc": "", "type": ["null", "File"],
"name": "clinicalRelevantVariants"}], "doc": ""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "SpecificDiseaseTitle",
        "clinicalRelevantTranscripts",
        "clinicalRelevantVariants",
        "geneIds",
        "relevantRegions",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'clinicalRelevantVariants': File,
            'relevantRegions': File,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'clinicalRelevantVariants': File,
            'relevantRegions': File,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'SpecificDiseaseTitle', 'clinicalRelevantTranscripts',
        'clinicalRelevantVariants', 'geneIds', 'relevantRegions'
    ]

    def __init__(self, **kwargs):
        self.SpecificDiseaseTitle = kwargs.get(
            'SpecificDiseaseTitle', None)
        self.clinicalRelevantTranscripts = kwargs.get(
            'clinicalRelevantTranscripts', None)
        self.clinicalRelevantVariants = kwargs.get(
            'clinicalRelevantVariants', None)
        self.geneIds = kwargs.get(
            'geneIds', None)
        self.relevantRegions = kwargs.get(
            'relevantRegions', None)


class VirtualPanels(ProtocolElement):
    """
    This record defines a gene panel. Geneneral and/or cancer specific
    """
    _schemaSource = """
{"namespace": "Gel_BioInf_Models", "type": "record", "name":
"VirtualPanels", "fields": [{"doc": "", "type": "string", "name":
"cancerName"}, {"doc": "", "type": "string", "name":
"dataModelCatalogueVersion"}, {"doc": "", "type": ["null", {"items":
"string", "type": "array"}], "name": "geneIds"}, {"doc": "", "type":
["null", {"items": "string", "type": "array"}], "name":
"clinicalRelevantTranscripts"}, {"doc": "", "type": ["null", {"doc":
"", "type": "record", "name": "File", "fields": [{"doc": "", "type":
["null", "string", {"items": "string", "type": "array"}], "name":
"SampleId"}, {"doc": "", "type": "string", "name": "URIFile"}]}],
"name": "RelevantRegions"}, {"doc": "", "type": ["null", "File"],
"name": "clinicalRelevantVariants"}], "doc": ""}
"""
    schema = avro.schema.parse(_schemaSource)
    requiredFields = {
        "RelevantRegions",
        "cancerName",
        "clinicalRelevantTranscripts",
        "clinicalRelevantVariants",
        "dataModelCatalogueVersion",
        "geneIds",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'RelevantRegions': File,
            'clinicalRelevantVariants': File,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'RelevantRegions': File,
            'clinicalRelevantVariants': File,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'RelevantRegions', 'cancerName',
        'clinicalRelevantTranscripts', 'clinicalRelevantVariants',
        'dataModelCatalogueVersion', 'geneIds'
    ]

    def __init__(self, **kwargs):
        self.RelevantRegions = kwargs.get(
            'RelevantRegions', None)
        self.cancerName = kwargs.get(
            'cancerName', None)
        self.clinicalRelevantTranscripts = kwargs.get(
            'clinicalRelevantTranscripts', None)
        self.clinicalRelevantVariants = kwargs.get(
            'clinicalRelevantVariants', None)
        self.dataModelCatalogueVersion = kwargs.get(
            'dataModelCatalogueVersion', None)
        self.geneIds = kwargs.get(
            'geneIds', None)
